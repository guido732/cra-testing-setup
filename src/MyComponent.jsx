// React
import React, { useState, useEffect } from "react";

export const MyComponent = ({
	error = false,
	success = false,
	disabled = false,
	size = "",
	fullWidth = false,
	value,
	id = "",
	label = "",
	type = "text",
	className = "",
	placeholder = "",
	onChange = () => {},
}) => {
	const [outputSize, setOutputSize] = useState("");
	const setSize = () =>
		fullWidth ? setOutputSize("100%") : size === "sm" ? setOutputSize("200px") : setOutputSize("400px");
	useEffect(setSize, []);

	return (
		<>
			<label className={`${disabled && "disabled"} ${error && "error"} ${success && "success"}`} htmlFor={id}>
				{label}
			</label>
			<input
				id={id}
				type={type}
				value={value}
				disabled={disabled}
				onChange={onChange}
				placeholder={placeholder}
				style={{ width: `${outputSize}` }}
				className={`${disabled && "disabled"} ${error && "error"} ${success && "success"} ${className}`}
			/>
		</>
	);
};
