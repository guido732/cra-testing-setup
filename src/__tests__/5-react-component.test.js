/**
 * Tests MyComponent Component
 *
 * @group demo5
 * @group demo
 */

// React
import React from "react";
// Testing Libs
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
// Component to Test
import { MyComponent } from "../MyComponent";

// Eliminar comentario para ver watch mode!
describe("MyComponent", () => {
	it("renders without crashing & matches snapshot", () => {
		const { asFragment } = render(<MyComponent label="demo-input" id="my-demo-input" />);
		expect(asFragment()).toMatchSnapshot();
		const mySelectedComponent = screen.getByRole("textbox", { name: /demo-input/i });
		expect(mySelectedComponent).toBeInTheDocument();
	});

	it("accepts an input and shows it", () => {
		render(<MyComponent label="demo-input" id="my-demo-input" />);
		const myText = "this is a demo text";
		const mySelectedComponent = screen.getByRole("textbox", { name: /demo-input/i });
		userEvent.type(mySelectedComponent, myText);
		expect(mySelectedComponent).toHaveValue(myText);
	});

	describe("state value", () => {
		it("is error", () => {
			render(<MyComponent label="demo-input" id="my-demo-input" error />);
			const mySelectedComponent = screen.getByRole("textbox", { name: /demo-input/i });
			expect(mySelectedComponent).toHaveClass("error");
			expect(mySelectedComponent).not.toHaveClass("success");
		});

		it("is success state", () => {
			render(<MyComponent label="demo-input" id="my-demo-input" success />);
			const myComponent = screen.getByRole("textbox", { name: /demo-input/i });
			expect(myComponent).toHaveClass("success");
			expect(myComponent).not.toHaveClass("error");
		});

		it("is disabled", () => {
			render(<MyComponent label="demo-input" id="my-demo-input" disabled />);
			const myComponent = screen.getByRole("textbox", { name: /demo-input/i });
			const myComponentTitle = screen.getByText("demo-input");
			expect(myComponent).toBeDisabled();
			expect(myComponent).toHaveClass("disabled");
			expect(myComponentTitle).toHaveClass("disabled");
		});
	});

	it("renders as a medium size (default/explicit)", () => {
		render(
			<>
				<MyComponent id="my-input-field-1" label="input Field 1" />
				<MyComponent id="my-input-field-2" label="input Field 2" size="md" />
			</>
		);
		const myComponent1 = screen.getByRole("textbox", { name: /input field 1/i });
		const myComponent2 = screen.getByRole("textbox", { name: /input field 2/i });
		expect(myComponent1).toHaveAttribute("style", "width: 400px;");
		expect(myComponent2).toHaveAttribute("style", "width: 400px;");
	});

	it("renders as fullwidth", () => {
		render(<MyComponent label="demo-input" id="my-demo-input" fullWidth />);
		const myComponent = screen.getByRole("textbox", { name: /demo-input/i });
		expect(myComponent).toHaveAttribute("style", "width: 100%;");
	});

	it("renders as small", () => {
		render(<MyComponent label="demo-input" id="my-demo-input" size="sm" />);
		const myComponent = screen.getByRole("textbox", { name: /demo-input/i });
		expect(myComponent).toHaveAttribute("style", "width: 200px;");
	});

	it("has a placeholder text visible", () => {
		const myText = "This is a placeholder text";
		render(<MyComponent label="demo-input" id="my-demo-input" placeholder={myText} />);
		const myComponent = screen.getByPlaceholderText(myText);
		expect(myComponent).toBeInTheDocument();
		expect(myComponent).toBeVisible();
	});

	it("doesn't have a placeholder text visible", () => {
		render(<MyComponent label="demo-input" id="my-demo-input" />);
		const myComponent = screen.getByRole("textbox", { name: /demo-input/i });
		expect(myComponent.placeholder).toBe("");
	});

	it("has a label visible", () => {
		render(<MyComponent label="demo-input" id="my-demo-input" />);
		const myComponent = screen.getByLabelText("demo-input");
		expect(myComponent).toBeInTheDocument();
		expect(myComponent).toBeVisible();
	});

	it("has a valid ID", () => {
		const myId = "my-input-field";
		render(<MyComponent label="demo-input" id={myId} />);
		const myComponent = screen.getByRole("textbox", { name: /demo-input/i });
		expect(myComponent).toHaveProperty("id", myId);
	});

	it("is an input of type Number", () => {
		render(<MyComponent label="demo-input" id="my-demo-input" type="number" />);
		const myComponent = screen.getByRole("spinbutton", { name: /demo-input/i });
		expect(myComponent).toHaveProperty("type", "number");
	});

	it("is an input of type Text", () => {
		render(<MyComponent label="demo-input" id="my-demo-input" type="text" />);
		const myComponent = screen.getByRole("textbox", { name: /demo-input/i });
		expect(myComponent).toHaveProperty("type", "text");
	});

	it("receives a className passed by props", () => {
		render(<MyComponent label="demo-input" id="my-demo-input" className="demo-classname" />);
		const myComponent = screen.getByRole("textbox", { name: /demo-input/i });
		expect(myComponent).toHaveClass("demo-classname");
	});
});
