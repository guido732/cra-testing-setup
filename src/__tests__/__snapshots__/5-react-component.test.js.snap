// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyComponent renders without crashing & matches snapshot 1`] = `
<DocumentFragment>
  <label
    class="false false false"
    for="my-demo-input"
  >
    demo-input
  </label>
  <input
    class="false false false "
    id="my-demo-input"
    placeholder=""
    style="width: 400px;"
    type="text"
    value=""
  />
</DocumentFragment>
`;
